dm/master/scheduler/scheduler.go

```diff
@@ -877,17 +877,32 @@ func (s *Scheduler) recoverWorkersBounds(cli *clientv3.Client) (int64, error) {
 		// set the stage as Free if it's keep alive.
 		if _, ok := kam[name]; ok {
 			w.ToFree()
-		}
-		// set the stage as Bound and record the bound relationship if exists.
-		if bound, ok := sbm[name]; ok {
-			err2 = s.updateStatusForBound(w, bound)
-			if err2 != nil {
-				return 0, err2
+			// set the stage as Bound and record the bound relationship if exists.
+			if bound, ok := sbm[name]; ok {
+				_, err2 := ha.PutSourceBound(cli, bound)
+				if err2 != nil {
+					return 0, err2
+				}
+				err2 = s.updateStatusForBound(w, bound)
+				if err2 != nil {
+					return 0, err2
+				}
+				delete(sbm, name)
 			}
 		}
 	}
 
-	// 5. recover bounds/unbounds, all sources which not in bounds should be in unbounds.
+	// 5. delete invalid source bound info in etcd
+	invalidSourceBounds := make([]string, 0, len(sbm))
+	for name := range sbm {
+		invalidSourceBounds = append(invalidSourceBounds, name)
+	}
+	_, err = ha.DeleteSourceBound(cli, invalidSourceBounds...)
```

Comment:
done in https://github.com/pingcap/dm/pull/483/commits/697b08fb1debf76dd3999e9a1998d1afc923ca5c

Reply:
I carefully think about the restart logic again. And I think it's better to trigger SourceBound in one transaction to avoid some workers are triggered while some are not. see more details in https://github.com/pingcap/dm/pull/483/commits/b5dae6b59971b614224276831e3913c587637d02